flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/GenericInMemoryCatalogStoreTest.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.configuration.Configuration;
+
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+/** Test for {@link GenericInMemoryCatalogStore}. */
+public class GenericInMemoryCatalogStoreTest {
+
+    @Test
+    void testStoreAndGet() {
+        CatalogStore catalogStore = new GenericInMemoryCatalogStore();
+        catalogStore.open();
+
+        catalogStore.storeCatalog(
+                "catalog1", CatalogDescriptor.of("catalog1", new Configuration()));
+        assertThat(catalogStore.getCatalog("catalog1").isPresent()).isTrue();
+        assertThat(catalogStore.contains("catalog1")).isTrue();
+
+        catalogStore.removeCatalog("catalog1", true);
+        assertThat(catalogStore.contains("catalog1")).isFalse();
+
+        catalogStore.close();
```

Comment:
Could we use `catalogStore` after it is closed? what's the expected behavior ?

Reply:
I think this may depend on the specific implementation of CatalogStoreFactory. 

In GenericInMemoryCatalogStoreFactory, it’s close methods do not handle any logic, just like InMemoryCatalog. So even after closing, you can still continue to use the getCatalog method. 

However, based on the current usage flow of CatalogStore, this situation is unlikely to occur. 

I also tried to create a CatalogStoreFactory that cannot continue to use getCatalog after closing, but I think it may be redundant and this process should be controlled by the specific implementation logic. Therefore, I did not add related tests in this PullRequest. 

What do you think?

Reply:
It's a little strange that `CatalogStore` can still be used after it is closed. Can we restrict this behavior? For example, `CatalogManager` should throw an exception when it access a closed `CatalogStore`

Reply:
IMO managing the state check of a `CatalogStore` outside of would not be an optimal approach, cause it would require anything that accesses a `CatalogStore` to do the same thing. I'd let the `CatalogStore` itself to manage its own state, so for the current scenario maybe we could initialize the in-memory map in the `open()` mehod, and clear/dereference it in the `close()`. We could store an `opened` or `closed` flag for it and based on that throw an exception for any access if the catalog store is not yet opened or closed already.

WDYT?

Reply:
Nice, how about introducing `AbstractCatalogStore` which implements `CatalogStore` to manage `opened` and `closed` flag in `open()` and `close()` methods? I think it should be universal for all `CatalogStore`.

Reply:
+1, that makes sense to me.

Reply:
This is indeed more general, but if we are only adding a flag, should we control the default behavior? For example, for `getCatalog`, should we check whether it is in a `closed` state or an `opened` state? Otherwise, simply adding a flag may not have much effect. However, in that case, we will need to add another method to `AbastractCatalogStore` for the actual implementation of `getCatalog`.

Reply:
I add the `AbastractCatalogStore` that store the open state and change the state when `open()` and `close`. 
And I add a `checkState` method for state checking,  and `isOpen` for getting state status. 
And now `GenericInMemoryCatalog` extends to `AbastractCatalogStore` and I add state check in  `getCatalog`, `listCatalogs`, `contains` , `removeCatalog`.  

WDYT? @ferenc-csaky @FangYongs @leonardBang

Reply:
+1